### DISCUSSION TITLE
TimeFormat relative time and reposition

### DISCUSSION THREAD

**Discussion by JRHB314** [AprilSylph/XKit-Rewritten#600] [Suggestion Box]
I saw it mentioned that since Tumblr has its own timestamps now, you'll probably discontinue the Timestamps feature, and just have the option to change the TimeFormat. That makes sense, but I have two requests if/when you do. 

The option to do relative time only, not just append. I was surprised to see this wasn't an option already, actually.

The other thing is I wish I could change the position of the timestamp to the top-right, the way it is now in New XKit and Xkit Rewritten timestamps. It just feels more cluttered having it directly under the usernames.

**Comment by EternityEngine** [comment]
I would also like to have the option of displaying only the relative time, as I find it easier to parse at a glance and also find the option of the entire date + relative time too long, visually. If it could also include the function where hovering over the relative time displays the proper time using the set custom format, as the Timestamps feature did, that would be fantastic.

It's odd to me that "relative time only" isn't an option, when it was in TimeFormat's predecessor.

**Reply by EternityEngine** [reply]
I've poked a bit at the Tumblr timestamps feature again and would also like to request the ability to move it from on the right of the "reblogged from" text in the header to underneath said text. An option to move the reblog item timestamps from underneath the blog name to the right side of the reblog item header wouldn't go amiss either.

Also, technically speaking, post timestamps do already have the "hover to display the full time" function, it's just blocked by the main and reblog item headers' post links (and of course does not follow the set custom timestamp format). Not sure if it's possible to fix that without moving those links (that would be its own tweak, I imagine; I do miss when the main post link was just in the corner of the post). Interestingly, with TimeFormat enabled, the cursor gets a question mark on it while hovering over a timestamp, which appears to be a deliberate CSS style choice in the feature despite there being no way to see it normally.

A caveat of the requested "relative time only" only option is that it _is_ technically the default Tumblr timestamp setup up to a certain time ago (it changes to "MMM D" at some point, and then "MMM D, YYYY"), it's just that Tumblr formats it in the extremely tiny "1h" way, instead of XKit's more readable "1 hour ago" way. It may make sense to have an option to format Tumblr's relative timestamp, or the relative timestamp TimeFormat adds, or some combination of the two.

**Reply by marcustyphoon** [reply]
Not sure we'd add that level of customization, but:

> would also like to request the ability to move it from on the right of the "reblogged from" text in the header to underneath said text.

Might want to `align-self` the blog avatar differently along with this.

```css
header .x66yu {
  flex-direction: column;
}
header .x66yu > .NB_7e {
  flex-direction: row;
}
header .x66yu > .NB_7e > .CtNGM {
  display: none;
}
```

> An option to move the reblog item timestamps from underneath the blog name to the right side of the reblog item header

Gonna be messy if a user has a lot of checkmarks, probably, but...

```css
._7Vla9 .l4Qpd > .cPL5a {
  flex-direction: row;
  align-items: center;
  justify-content: start;
  gap: 1ch;
}
```

> post timestamps do already have the "hover to display the full time" function, it's just blocked by the main and reblog item headers' post links

I have a PR to fix this one!

> Interestingly, with TimeFormat enabled, the cursor gets a question mark on it while hovering over a timestamp, which appears to be a deliberate CSS style choice in the feature despite there being no way to see it normally.

Yeah, it used to work a while ago before that link was added.

**Reply by EternityEngine** [reply]
Thank you for the CSS suggestions! The second one isn't quite what I was asking for, but that's my fault for not specifying - I meant the far right, next to the meatball menu. Replacing `justify-content: start;` with `justify-content: space-between;` gets it there. (Also the `gap` seems unnecessary?)

Unfortunately they don't work with the Old Tumblr Dashboard userstyle I'm using - the first one has the "reblogged" text and timestamp just stack on top of each other next to the username (which is why I was hoping for an XKit solution, as that's all I could get as well), and the second one just straight up does nothing, though I did figure out my own solution there. Maybe I'll figure out a userscript solution for the main timestamp sometime in the future...

To move the reblog item timestamps to the far right with [Old Tumblr Dashboard](https://userstyles.world/style/11286/) installed:
```
#base-container > div.D5eCV > .c9lq9 > div._3xgk.Ril26 > .gPQR5.FGfuE.iTX1z > div.lSyOz > .O4V_R > div.Evcyl > div > div > div > div > div > article > div > div.LaNUG > div > div > span > div > div._7Vla9 > div {
  justify-items: normal;
  justify-content: unset;
} /* allow reblog item timestamp repositioning */
@media (min-width: 990px) {
  .u2tXn .m2Bdx.pLL0g.PCzZt > .l4Qpd > .cPL5a {justify-content: space-between;} /* reblog item header */
  .u2tXn .x66yu .QOpFt {margin-right: 30px; --timestampfont: .787rem;} /* reblog item timestamp */
}
```

(I wish it could be less ridiculously precise, but it has to be to override the userstyle without using `!important` and potentially overriding CSS further in.)

**Reply by marcustyphoon** [reply]
Yeah, I'm not sure how pixiel generates the selectors for that style—possibly using the ublock origin tool, if I had to guess? I think they used to be more handwritten, but with the sheer amount of work required to keep a style like that updated, I'm not surprised if making use of tools is the way to go.

What did you want to do with the first one, again? The old tumblr dashboard style already puts the timestamp on a separate line from the "reblogged" text.

**Reply by EternityEngine** [reply]
> What did you want to do with the first one, again? The old tumblr dashboard style already puts the timestamp on a separate line from the "reblogged" text.

It actually doesn't. What the Old Tumblr Dashboard style does is move the avatar out of the main post header back to the left of the post and move the line of "Reblogged [blogname] · [timestamp]" up to the same line as the poster name, so it's all on one line, i.e. "[Poster] reblogged [blogname] · [timestamp]". I want to move the "[timestamp]" part down to underneath the "[Poster] reblogged [blogname]" part, to emulate how XKit's Timestamps feature worked, as it feels more readable that way, especially with the longer timestamps using TimeFormat. Unfortunately, it doesn't seem possible to do in tandem with the all-one-line change with just CSS, due to how the HTML is nested, since "Reblogged [blogname] · [timestamp]" is all one element on the same level as "[Poster]". So the current header/timestamp CSS layout options appear to be:

Tumblr default:
> [Poster]
> Reblogged [blogname] · [timestamp]

Old Tumblr Dashboard:
> [Poster] reblogged [blogname] · [timestamp]

Theoretical option:
> [Poster]
> Reblogged [blogname]
> [timestamp]

What happens when trying to separate "reblogged [blogname]" and "[timestamp]" with Old Tumblr Dashboard installed:

<img width="357" height="49" alt="image" src="https://github.com/user-attachments/assets/2e8f05ce-5c02-49b7-9a4e-d66f409d8487" />

When what I was hoping for was:

> [Poster] reblogged [blogname]
> [timestamp]

Which is unfortunate, but unless there is some obscure CSS wizardry I am unaware of that can get around this, also unfixable without JavaScript.

**Reply by marcustyphoon** [reply]
What's your Old Tumblr Dashboard version number? Mine is 17.16 and I'm seeing what you were hoping for.

**Reply by EternityEngine** [reply]
That's odd. I'm also on 17.16 and it's all on one line for me. I even tried reinstalling it, and installing it with Stylus in case my use of xStyle was the culprit, to no avail. I only get timestamps on the second line if I enable Timestamps in New XKit.

<img width="477" height="130" alt="image" src="https://github.com/user-attachments/assets/28d3d3b1-d5bd-413c-9309-89f861427c90" />

(This is with none of my custom CSS enabled, so that's not the culprit either.)

**Reply by marcustyphoon** [reply]
I think the relevant Old Tumblr Dashboard css rule is this one, which... uh, let's be real, there could be any number of reasons why that didn't hit. Maybe try making it less specific?

```css
#base-container > div.D5eCV > div > div._3xgk.Ril26 > div > div.lSyOz > main > div.Evcyl > div.zAlrA > div > div > div > div > article > div > header > div > div > div.l4Qpd > div.cPL5a.QCWsQ.zmkgk > div.x66yu > div.QOpFt {
  position: absolute;
  top: 46px;
  left: 19px;
  margin-top: -6px;
}
```

**Reply by EternityEngine** [reply]
Going through the HTML in the inspector, it looks like `div.zAlrA` is the culprit (the immediate child of `div.Evcyl` does not have a class assigned, at least for me). Removing `.zAlrA` makes it work, although the following works just as well:

```
header .x66yu > .QOpFt {
  position: absolute;
  top: 40px;
  left: 19px;
}
```

Or, to align it with the XKit timestamp (and expand the header to fit):

```
header .x66yu > .QOpFt {
  position: absolute;
  top: 47px;
  left: 16px;
}
.DEkbl {
  padding-bottom: 33px;
}
```

As a bonus, I already had CSS to shrink the post header link so it didn't overlap the XKit timestamp, so now the hover function works for that Tumblr timestamp.

```
.DEkbl .mYcfU {
  height: 44px;
}
```

Of course, just now I had a thought that changing the z-index property of the timestamps would fix the hover function, and in testing, setting their z-index to 0 or higher allows it to work. Your PR fix looks more fit-for-purpose, though.

**Reply by EternityEngine** [reply]
Coming back just to say that another reason to have an XKit solution to moving the main post timestamp (or at least decoupling it from the "[poster] reblogged" header text) is that using `position: absolute` to reposition it means that anytime something else is added to the header (e.g. Content Labels, "Pinned" post label), the timestamp then overlaps with the poster header due to the added element bumping the poster header downwards. It's a minor irk, and fixable with extra CSS (fortunately those two examples share a class), but still.

I've since had to revamp the CSS for my particular pair of requests due to Tumblr changing things slightly, Old Tumblr Dashboard revisions, noticing problems with search results, and the aforementioned issue, so I figured I could post the updated CSS here as well.

To move the main post timestamp to underneath the "reblogged from" text, with Old Tumblr Dashboard installed (aligned with XKit Timestamps):
```
header:not(.rtjhW) .x66yu > .NB_7e > .CtNGM {
  display: none;
} /* dot before header timestamp */
header:not(.rtjhW) .x66yu > .QOpFt {
  position: absolute;
  top: 45px;
  left: 16px;
} /* header timestamp */
header:not(.rtjhW) .Pl0ah + .m2Bdx .x66yu > .QOpFt {
  position: absolute;
  top: 66px;
  left: 16px;
} /* header timestamp after content label/pinned post */
.DEkbl {
  padding-bottom: 33px;
} /* post header */
```
To move the reblog item timestamps to the right end of the reblog item header, with Old Tumblr Dashboard installed:
```
#root div div#base-container div.D5eCV div.c9lq9 div._3xgk.VsR8V.f_ziq.Ril26 div.gPQR5.pTaNm div.lSyOz main.UX61Y div.Evcyl.W0qvx div.hyiL2 div.USr9U div div.So6RQ.i4aRn.YSitt div.ge_yK div.c79Av article.FtjPK div.LaNUG div div.Qb2zX span div.u2tXn div._7Vla9 div.cPL5a.Afu4V.QCWsQ {
  flex-direction: column;
  align-items: unset;
} /* allow grid search results reblog item timestamp repositioning */
#root div div#base-container div.D5eCV div.c9lq9 div._3xgk.VsR8V.f_ziq.Ril26 div.gPQR5.pTaNm div.lSyOz main.UX61Y div.Evcyl.W0qvx div.hyiL2 div.USr9U div div.So6RQ.i4aRn.YSitt div.ge_yK div.c79Av article.FtjPK div.LaNUG div div.Qb2zX span div.u2tXn div._7Vla9 div.cPL5a.Afu4V.QCWsQ div.m2Bdx.pLL0g.PCzZt {
  max-width: 100%;
} /* allow grid search results reblog item timestamp full width */
.u2tXn .pLL0g.PCzZt > .l4Qpd > .cPL5a {
  justify-content: space-between;
} /* reblog item header */
```
I do also have CSS to hide TimeFormat's absolute timestamps (so, only display the relative timestamps), and may as well share it (I don't hide it for reblog notes personally, but I went and got it for completeness' sake):
```
.QOpFt [data-formatted-time]::before, .X9zTC [data-formatted-time]::before, .zWtFe [data-formatted-time]::before {
  display: none;
} /* header & reblog item, reply notes, & reblog notes absolute timestamps */
```
Please note that if "Append relative time" is not enabled in TimeFormat's settings, this will just hide the timestamps entirely.

**Comment by marcustyphoon** [comment]
Hm—I guess putting nothing in the format box doesn't do this, because it just makes the default format, huh. Seems like a reasonable feature request.
